### PR TITLE
True TOML config support

### LINKED ISSUE #999: native pyproject.toml config
This is a ticket to track and implement a native ``pyproject.toml`` configuration. Will update with the proposed plan and format.

Core goals:
- explicitly inheritance based (a lot of substitution at the moment is used to replicate inheritance), 
- substitution able,
- reasonably familiar (e.g. share most concepts with ini format),
- succinct (some form of factors needs to have).

We'll use this opportunity to rename some keys (e.g. ``skipsdist`` should be ``skip_packaging``, etc).

### PR DESCRIPTION
**Pull request by gaborbernat**
Done:
- `pyproject.toml` source
- `tox.toml` source
- toml loader with type validations
- variable substitution that makes sense for TOML files
- reference mechanism that makes sense for TOML files
- documentation.

https://tox--3353.org.readthedocs.build/en/3353/config.html#tox-toml

Resolves #999.

### CODE REVIEW

**Comment by gaborbernat** [comment]
@ssbarnea @ziima this is now ready for first release, give it a go if you can.

File: docs/installation.rst [hunk 1]
```diff
@@ -1,31 +1,42 @@
...
 
-via pipx
+As tool
 --------
 
-:pypi:`tox` is a CLI tool that needs a Python interpreter (version 3.7 or higher) to run. We recommend :pypi:`pipx` to
-install tox into an isolated environment. This has the added benefit that later you'll be able to upgrade tox without
-affecting other parts of the system.
+:pypi:`tox` is a CLI tool that needs a Python interpreter (version 3.7 or higher) to run. We recommend either
+:pypi:`pipx` or  :pypi:`ruff` to install tox into an isolated environment. This has the added benefit that later you'll
```
Comment (cthoyt) [inline docs/installation.rst:8]: is this a typo? should it be uv and not ruff?

Comment (gaborbernat) [reply docs/installation.rst:8]: Ah yes!

File: docs/config.rst [hunk 2]
```diff
@@ -9,33 +9,47 @@ be set once and used for all tox environments, while environment options are app
...
+1. ``tox.ini`` (INI),
+2. ``setup.cfg`` (INI),
+3. ``pyproject.toml`` with the ``tool.tox`` table, having ``legacy_tox_ini`` key (containing INI),
+4. Native ``pyproject.toml`` under the ``tool.tox`` table (TOML),
+5. ``tox.toml`` (TOML).
+
+Historically, the INI format was created first, and TOML was added in 2024. The TOML format generally is more robust,
+however is less powerful and more verbose. You should use TOML unless you need some of the more advanced features
+that TOML does not support (such as conditional factors, generative environments to name a few -- however, PRs to
+add support for these in TOML is welcome).
```
Comment (ziima) [inline docs/config.rst:28]: I'd mark TOML config as experimental until all these are supported. Supporting those may affect the TOML file structure.

File: docs/config.rst [hunk 3]
```diff
@@ -97,6 +113,58 @@ examples first and use this page as a reference.
...
+
+    [tool.tox.env.type]
+    description = "run type check on code base"
+    deps = ["mypy==1.11.2", "types-cachetools>=5.5.0.20240820", "types-chardet>=5.0.4.6"]
+    commands = [["mypy", "src{/}tox"], ["mypy", "tests"]]
+
+``tox.toml``
+~~~~~~~~~~~~
+
+This configuration file is equivalent to :ref:`pyproject.toml - native <pyproject-toml-native>` with the difference
```
Comment (ziima) [inline docs/config.rst:149]: I very appreciate the option to have a separate TOML file :+1:

File: docs/user_guide.rst [hunk 4]
```diff
@@ -18,57 +18,98 @@ Configuration
...
+
+
+    .. tab:: INI
+
+       .. code-block:: ini
+
+           [tox]
+            requires =
+                tox>=4
+            env_list = lint, type, 3.{3,2,1}
```
Comment (ziima) [inline docs/user_guide.rst:57]:
```
env_list = lint, type, 3.{13,12,11}
```
To make it correspond to the TOML version.

File: docs/user_guide.rst [hunk 5]
```diff
@@ -18,57 +18,98 @@ Configuration
...
+          env_list = ["lint", "type", "3.13", "3.12", "3.11"]
+
+    .. tab:: INI
+
+       .. code-block:: ini
+
+        [tox]
+        requires =
+            tox>=4
+        env_list = lint, type, 3.{3,2,1}
```
Comment (ziima) [inline docs/user_guide.rst:111]:
```
env_list = lint, type, 3.{13,12,11}
```

File: docs/config.rst [hunk 6]
```diff
@@ -1019,116 +1130,243 @@ changed via ``TOX_USER_CONFIG_FILE`` environment variable. Example configuration
...
+       tox --override testenv.ignore_errors=True
+
+
+You could add additional dependencies by running:
+
+.. tab:: TOML
+
+    .. code-block:: bash
+
+       tox --override env_run_base.deps+=pytest-xdist
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply docs/config.rst:1220]: Achieving this goal would require that the syntax for the new file format to be very ugly. testenv:whatever Is very much something we inherited in the INI world. Furthermore, this is already the case when using setup.CFG over tox.ini.

File: docs/config.rst (hunk 2 above)
Comment (gaborbernat) [reply docs/config.rst:28]: I disagree. Some of these features I explicitly do not want to support in the new format (such as conditional factors). The TOML file Format is just not a good match for it.

File: docs/user_guide.rst (hunk 4 above)
Comment (gaborbernat) [reply docs/user_guide.rst:57]: Ah good typo spot.

File: docs/config.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply docs/config.rst:1220]: If you read the configuration file format introduction section, it clearly calls out the differences.

**Pull request merged**